pkg/storage/cacher.go

```diff
@@ -402,10 +403,20 @@ func filterFunction(key string, keyFunc func(runtime.Object) (string, error), fi
 			glog.Errorf("invalid object for filter: %v", obj)
 			return false
 		}
-		if !strings.HasPrefix(objKey, key) {
-			return false
+
+		segments := strings.Split(objKey, "/")
```

Comment:
Do you mean:
```
if !strings.HasPrefix(objKey, key+"/") {
        return false
}
```

I don't think it's correct if client watch a key(not a dir). For example, if client watch "/pods/ns1/pod1", cache will also return events about "/pods/ns1/pod11", since key "ns1/pod1" is also prefix of objKey "ns1/pod11"+"/"